src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1114 @@
+    }
+
+    const { inputs, fields, mutation, blockData: inputBlockData } = serializeInputsResult;
+
+    applyBlockData(blockData, inputBlockData);
+
+    const obj: sb3.Block = {
+      opcode: block.opcode,
+
+      parent: parent ? parent.id : null,
+      next: nextBlockId,
+      topLevel: !parent,
+
+      inputs,
+      fields,
+      mutation,
+
+      shadow: false
+    };
+
+    if (obj.topLevel) {
+      obj.x = options.x;
+      obj.y = options.y;
+    }
+
+    const blockId = block.id;
+
+    blockData[blockId] = obj;
+
+    return { blockData, blockId };
+  }
+
+  interface CustomBlockData {
+    args: Array<{
+      default: string;
+      id: string;
+      name: string;
+      type: "boolean" | "numberOrString";
+    }>;
+    warp: boolean;
+  }
+
+  type GetCustomBlockData = (proccode: string) => CustomBlockData;
+
+  type BaseGetBroadcastId = (name: string) => string;
+  interface GetBroadcastId extends BaseGetBroadcastId {
+    initialBroadcastName: string;
+  }
+
+  function collectCustomBlockData(target: Target): GetCustomBlockData {
+    // Parse the scripts in a target, collecting metadata about each custom
+    // block's arguments and other info, and return a function for accessing
+    // the data associated with a particular custom block's proccode.
+    //
+    // It's necesary to collect this data prior to serializing any associated
+    // procedure_call blocks, because they require access to data only found
+    // on the associated procedure_definition. (Specifically, the types of
+    // each input on the custom block, since those will influence the initial
+    // value & shadow type in the serialized caller block's inputs.)
+
+    const data: { [proccode: string]: CustomBlockData } = {};
+
+    for (const script of target.scripts) {
+      const block = script.blocks[0];
+      if (block.opcode !== OpCode.procedures_definition) {
+        continue;
+      }
+
+      const proccode = block.inputs.PROCCODE.value;
+      const warp = block.inputs.WARP.value;
+
+      const args: Array<{
+        default: string;
+        id: string;
+        name: string;
+        type: "boolean" | "numberOrString";
+      }> = [];
+
+      for (const { name, type } of block.inputs.ARGUMENTS.value) {
+        if (type === "label") {
+          continue;
+        }
+
+        const id = generateId();
+
+        args.push({
+          id,
+          name,
+          type,
+          default: prop(
+            {
+              boolean: "false",
+              numberOrString: ""
+            },
+            type
+          )
+        });
       }
-      return { ...variables, [variable.id]: varData };
-    }, {}),
-    lists: project.stage.lists.reduce(
-      (lists: sb3.Stage["lists"], list) => ({
-        ...lists,
-        [list.id]: [list.name, list.value]
+
+      data[proccode] = { args, warp };
+    }
+
+    return (proccode: string): CustomBlockData => {
+      return data[proccode];
+    };
```

Comment:
Not entirely sure I agree with the design pattern of dynamically creating getter functions.

Yes, the data structure is now "opaque", but IMO it makes the code much harder to reason about. Instead of taking one *specific* type of block data, `serializeBlock` could theoretically take *any* function that returns `CustomBlockData` in any number of ways. You have to trawl through the code just to figure out that there's only one concrete implementation called in one place.